src/lib/homestore_backend/pg_blob_iterator.cpp

```diff
@@ -36,102 +84,169 @@ PG* HSHomeObject::PGBlobIterator::get_pg_metadata() {
     return iter->second.get();
 }
 
-void HSHomeObject::PGBlobIterator::create_pg_shard_snapshot_data(sisl::io_blob_safe& meta_blob) {
-    // auto pg = get_pg_metadata();
-    // auto& pg_info = pg->pg_info_;
-    // auto& pg_shards = pg->shards_;
-    //
-    // flatbuffers::FlatBufferBuilder builder;
-    // std::vector< std::uint8_t > uuid(pg_info.replica_set_uuid.size());
-    // std::copy(pg_info.replica_set_uuid.begin(), pg_info.replica_set_uuid.end(), uuid.begin());
-    // auto pg_entry = CreatePGInfoEntry(builder, pg_info.id, 0 /* priority*/, builder.CreateVector(uuid));
-    //
-    // std::vector< ::flatbuffers::Offset< ShardInfoEntry > > shard_entries;
-    // for (auto& shard : pg_shards) {
-    //     auto& shard_info = shard->info;
-    //     // TODO add lsn.
-    //     shard_entries.push_back(CreateShardInfoEntry(
-    //         builder, static_cast< uint8_t >(shard_info.state), shard_info.placement_group, shard_info.id,
-    //         shard_info.total_capacity_bytes, shard_info.created_time, shard_info.last_modified_time));
-    // }
-    // builder.FinishSizePrefixed(CreateResyncPGShardInfo(builder, pg_entry, builder.CreateVector(shard_entries)));
-    // meta_blob = sisl::io_blob_safe{builder.GetSize()};
-    // std::memcpy(meta_blob.bytes(), builder.GetBufferPointer(), builder.GetSize());
+bool HSHomeObject::PGBlobIterator::create_pg_snapshot_data(sisl::io_blob_safe& meta_blob) {
+    auto pg = home_obj_._pg_map[pg_id_].get();
+    if (pg == nullptr) {
+        LOGE("PG not found in pg_map, pg_id={}", pg_id_);
+        return false;
+    }
+    auto pg_info=pg->pg_info_;
+
+    flatbuffers::FlatBufferBuilder builder;
+    std::vector< std::uint8_t > uuid(pg_info.replica_set_uuid.begin(), pg_info.replica_set_uuid.end());
+
+    std::vector<::flatbuffers::Offset<homeobject::Member>> members(pg_info.members.size());
+    for (auto& member : pg_info.members) {
+        auto id=std::vector< std::uint8_t >(member.id.begin(), member.id.end());
+        members.push_back(CreateMemberDirect(builder, &id, member.name.c_str(), member.priority));
+    }
+    std::vector<uint64_t> shard_ids;
+    for (auto& shard : pg->shards_) {
+        shard_ids.push_back(shard->info.id);
+    }
+
+    auto pg_entry = CreateResyncPGMetaDataDirect(builder, pg_info.id, &uuid, pg->durable_entities().blob_sequence_num, pg->shard_sequence_num_ , &members, &shard_ids);
+    builder.FinishSizePrefixed(pg_entry);
+    auto payload = sisl::io_blob_safe{builder.GetSize()};
+    std::memcpy(payload.bytes(), builder.GetBufferPointer(), builder.GetSize());
+
+    pack_resync_message(meta_blob, SyncMessageType::PG_META, payload);
+    return true;
 }
 
-void HSHomeObject::PGBlobIterator::create_pg_snapshot_data(sisl::io_blob_safe& meta_blob) {
-    //TODO
+bool HSHomeObject::PGBlobIterator::generate_shard_blob_list() {
+    auto cur_shard_seq_num=cur_shard_id & 0xFFFF;
```

Comment:
Should be `0xFFFFFFFFFFFF` since the seq_num occupies 48 bits within shard ID?

And here's a problem throughout this file regarding the seq_num and full shard_id: the `shard_id` field in objId actually represents the shard seq_num, so we cannot directly compare it with shard_ids in the shard_list. (Perhaps we should renaming the field in objId to avoid confusion...)

Reply:
Good point, let's rename objId.shard_id to objId.shard_seq_num to distinguish them.